epf/Cell.cpp

```diff
@@ -62,33 +54,46 @@ void Cell::advance()
 CellMgr::CellMgr(int pointSize, Writer *writer) : m_pointSize(pointSize), m_writer(writer)
 {}
 
-
-Cell *CellMgr::get(const VoxelKey& key)
+Cell *CellMgr::get(const VoxelKey& key, const Cell *lastCell)
 {
     auto it = m_cells.find(key);
     if (it == m_cells.end())
     {
-        Cell::FlushFunc f = [this](Cell *exclude)
-        {
-            flush(exclude);
-        };
-        std::unique_ptr<Cell> cell(new Cell(key, m_pointSize, m_writer, f));
+        std::unique_ptr<Cell> cell(new Cell(key, m_pointSize, m_writer, this, lastCell));
         it = m_cells.insert( {key, std::move(cell)} ).first;
     }
-    Cell& c = *(it->second);
-    return &c;
+
+    return it->second.get();
+}
+
+DataVecPtr CellMgr::getBuffer(const Cell *exclude)
+{
+    DataVecPtr b = m_writer->fetchBuffer();
+
+    // If we couldn't fetch a buffer, flush all the the buffers for this processor and
+    // try again, but block.
+    if (!b)
+    {
+        flush(exclude);
+        b = m_writer->fetchBufferBlocking();
+    }
+    return b;
 }
 
 // Eliminate all the cells and their associated data buffers except the `exclude`
 // cell.
-void CellMgr::flush(Cell *exclude)
+void CellMgr::flush(const Cell *exclude)
 {
     CellMap::iterator it = m_cells.end();
     if (exclude)
+    {
         it = m_cells.find(exclude->key());
+        assert(it != m_cells.end());
```

Comment:
should this be a real throw? I assume this is a fatal condition, correct?

Reply:
I'll remove it. It was mostly there for me during testing.